File: src/main/java/com/micronaut/bug/config/ObservationConfig.java
package com.micronaut.bug.config;

import io.micrometer.context.ContextRegistry;
import jakarta.annotation.PostConstruct;
import org.slf4j.MDC;
import org.springframework.boot.web.embedded.netty.NettyServerCustomizer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import reactor.core.publisher.Hooks;
import reactor.netty.http.HttpProtocol;

@Configuration
public class ObservationConfig {

    public static final String X_REQ_ID = "x-req-id";
    public static final String TARGET_ID = "targetId";

    @PostConstruct
    public void init() {
        // 1. Включаем автоматический проброс в Reactor
        Hooks.enableAutomaticContextPropagation();

        var registry = ContextRegistry.getInstance();

        // 2. Регистрируем наш кастомный ключ x-req-id
        registry.registerThreadLocalAccessor(
            X_REQ_ID,
            LoggingContext::get,
            LoggingContext::set,
            LoggingContext::reset
        );

        registry.registerThreadLocalAccessor(TARGET_ID,
            () -> MDC.get(TARGET_ID),
            val -> MDC.put(TARGET_ID, val),
            () -> MDC.remove(TARGET_ID)
        );

        registry.registerThreadLocalAccessor(
            "current-user",
            UserContext::get,
            UserContext::set, // Используем временный ThreadLocal как транспорт
            UserContext::reset
        );
    }

//    @Bean
//    public NettyServerCustomizer http2CleartextCustomizer() {
//        return server -> server.protocol(HttpProtocol.H2C)
//            .wiretap(true);
//    }
}

File: src/main/kotlin/com/micronaut/bug/config/ObservationConfig.kt
package com.micronaut.bug.config

import com.micronaut.bug.config.ServerLoggingFilter.Companion.MDC_RQ_ID
import com.micronaut.bug.config.log.MdcConverter.Companion.MDC_TARGET_ID
import io.micrometer.context.ContextRegistry
import org.slf4j.MDC

//@Configuration
class ObservationConfig {

    init {
        val registry = ContextRegistry.getInstance()

        // Регистрируем кастомный ключ x-req-id
        registry.registerThreadLocalAccessor(
            MDC_RQ_ID,
            { MDC.get(MDC_RQ_ID) },
            { value -> MDC.put(MDC_RQ_ID, value) },
            { MDC.remove(MDC_RQ_ID) }
        )

        // Регистрируем кастомный ключ targetId
        registry.registerThreadLocalAccessor(
            MDC_TARGET_ID,
            { MDC.get(MDC_TARGET_ID) },
            { value -> MDC.put(MDC_TARGET_ID, value) },
            { MDC.remove(MDC_TARGET_ID) }
        )
    }
}
